contracts/StakingVe.sol

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+/// @title A vesting style staking contract with extendable linear decay
+/// @author Auroter
+/// @notice Allows you to lock tokens in exchange for governance tokens
+/// @notice Locks can be extended or deposited into
+/// @notice Maximum deposit duration is two years (104 weeks)
+/// @dev Simply call stake(...) to create initial lock or extend one that already exists for the user
+
+// We have to maintain old versions of this to use legacy solidity (0.6.12)
+import "./libraries/SafeERC20.sol";
+import "./libraries/SafeMath.sol";
+import "./interfaces/IERC20.sol";
+
+//import "hardhat/console.sol";
+
+contract StakingVe {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    // Info pertaining to staking contract
+    address private StakedToken; // An ERC20 Token to be staked (i.e. Axial)
+    address private Governance; // Who has administrative access
+    string private Name; // New asset after staking (i.e. sAxial)
+    string private Symbol; // New asset symbol after staking (i.e. sAXIAL)
+    uint256 private InterpolationGranularity = 1e18; // Note: ERC20.decimals() is for display and does not affect arithmetic!
+
+    // Info pertaining to users
+    address[] private Users; // An array containing all user addresses
+    mapping(address => LockVe) private Locks; // A mapping of each users lock
+    mapping(address => uint256) private LockedFunds; // A mapping of each users total deposited funds
+
+    // Lock structure, only one of these is allowed per user
+    // A DELTA can be derived as the degree of interpolation between the start/end block:
+    // Delta = (end - now) / end - start
+    // This can be used to determine how much of our staked token is unlocked:
+    // currentAmountLocked = startingAmountLocked - (delta * startingAmountLocked)
+    struct LockVe {
+        uint256 StartBlockTime;
+        uint256 EndBlockTime;
+        uint256 StartingAmountLocked;
+        bool Initialized;
+    }
+
+    uint8 internal constant _not_entered = 1;
+    uint8 internal constant _entered = 2;
+    uint8 internal _entered_state = 1;
+    /// @notice Protects deposit / withdraw functions from exploits
+    /// @notice Ensures a function cannot be executed again while it is still running
+    /// @dev TODO: Move this into its own contract
+    modifier nonreentrant() {
```

Comment:
We could inherit from OpenZeppelins ReentrancyGuard contract for this instead of writing our own.